File: python/examples/core/05-streaming/main.py
"""Example 05: subscribe to a canvas's notes endpoint for a fixed duration.

Opens the SDK's NDJSON streaming subscription, prints a one-line summary per
event, and exits cleanly after ``STREAM_DURATION_SECONDS`` (default 30) or on
``SIGINT``.

The Canvus streaming protocol sends:
    - a snapshot of all currently-existing items first;
    - then incremental updates as JSON objects (each newline-terminated);
    - and occasional empty keep-alive lines that the SDK silently drops.
"""

from __future__ import annotations

import asyncio
import contextlib
import os
import signal
import sys
from typing import Any

import structlog
from pydantic import Field, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict

from canvus_sdk import APIError, Client, configure_logging

logger = structlog.get_logger(__name__)


class StreamSettings(BaseSettings):
    """Environment configuration for the streaming example."""

    model_config = SettingsConfigDict(
        env_prefix="CANVUS_",
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    api_url: str = Field(...)
    api_key: str = Field(...)
    canvas_id: str = Field(...)


def _stream_duration() -> float:
    raw = os.environ.get("STREAM_DURATION_SECONDS", "30")
    try:
        return float(raw)
    except ValueError:
        logger.warning("STREAM_DURATION_SECONDS invalid; using 30", raw=raw)
        return 30.0


async def _consume(
    client: Client,
    canvas_id: str,
    stop_event: asyncio.Event,
) -> tuple[int, int]:
    """Drain the subscription until ``stop_event`` is set.

    Returns (event_count, batch_count).
    """
    events_total = 0
    batches = 0
    sub = client.widgets.subscribe(canvas_id, widget_type="notes")
    # `subscribe()` returns an async generator directly (it `yield`s); we
    # iterate it under a race with the stop event so SIGINT or the duration
    # timeout cuts in promptly.
    aiter = sub.__aiter__()
    while not stop_event.is_set():
        next_task = asyncio.create_task(aiter.__anext__())
        stop_task = asyncio.create_task(stop_event.wait())
        done, pending = await asyncio.wait(
            {next_task, stop_task},
            return_when=asyncio.FIRST_COMPLETED,
        )
        for p in pending:
            p.cancel()
        if next_task in done:
            try:
                payload: Any = next_task.result()
            except StopAsyncIteration:
                logger.info("stream ended by server")
                break
            except APIError as e:
                logger.error(
                    "stream api error",
                    status_code=e.status_code,
                    response_body=e.response_body,
                )
                break
            batches += 1
            if isinstance(payload, list):
                events_total += len(payload)
                first = payload[0] if payload else None
                logger.info(
                    "batch",
                    batch=batches,
                    count=len(payload),
                    first_id=first.get("id") if isinstance(first, dict) else None,
                    first_text=(
                        first.get("text", "")[:60] if isinstance(first, dict) else None
                    ),
                )
            elif isinstance(payload, dict):
                events_total += 1
                logger.info(
                    "event",
                    batch=batches,
                    id=payload.get("id"),
                    text=payload.get("text", "")[:60],
                )
            else:
                logger.debug("unknown payload type", payload_type=type(payload).__name__)
        else:
            # stop_event won the race
            break
    return events_total, batches


async def main() -> int:
    """Entry point."""
    configure_logging()
    try:
        settings = StreamSettings()  # type: ignore[call-arg]
    except ValidationError as e:
        logger.error("config invalid", errors=e.errors())
        return 2

    duration = _stream_duration()
    stop_event = asyncio.Event()
    loop = asyncio.get_running_loop()

    def _signal_handler() -> None:
        logger.info("sigint received; draining stream")
        stop_event.set()

    # Windows event loop doesn't support add_signal_handler.
    for sig in (signal.SIGINT, signal.SIGTERM):
        with contextlib.suppress(NotImplementedError):
            loop.add_signal_handler(sig, _signal_handler)

    async with Client(settings.api_url, settings.api_key) as client:
        logger.info(
            "subscribing",
            canvas_id=settings.canvas_id,
            duration_seconds=duration,
        )
        consumer = asyncio.create_task(
            _consume(client, settings.canvas_id, stop_event),
        )
        try:
            await asyncio.wait_for(asyncio.shield(consumer), timeout=duration)
        except TimeoutError:
            logger.info("duration elapsed; signalling stop")
            stop_event.set()
            events_total, batches = await consumer
        else:
            events_total, batches = consumer.result()

    logger.info(
        "stream finished",
        batches=batches,
        events_total=events_total,
        duration_seconds=duration,
    )
    print(f"received {events_total} events across {batches} batches in {duration}s")
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: go/examples/core/05-streaming/main.go
// Command 05-streaming subscribes to the notes endpoint of a Canvus canvas
// and prints summary lines for each NDJSON event the server emits, exiting
// cleanly on SIGINT or after STREAM_DURATION_SECONDS elapses.
//
// The Canvus SDK does not yet expose a typed Subscribe helper, so this
// example builds the HTTP request directly using the Session's exported
// HTTPClient field — the API-key round-tripper installed by WithAPIKey is
// still in effect, so authentication "just works" without re-implementing
// the header logic.
//
// Wire shape: GET /api/v1/canvases/{id}/notes?subscribe=true returns a
// streaming response of newline-delimited JSON. Each frame is the full
// current list of notes, not a delta — clients receive the latest snapshot
// on connect plus a fresh snapshot every time something changes. Empty
// lines are keep-alives and must be ignored.
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"path"
	"strconv"
	"sync/atomic"
	"syscall"
	"time"

	"github.com/jaypaulb/MT-Canvus-Tools/go/sdk/canvus"
)

const defaultDurationSeconds = 30

func main() {
	setupLogging()
	if err := run(); err != nil {
		slog.Error("streaming failed", "err", err)
		os.Exit(1)
	}
}

func run() error {
	baseURL, err := mustEnv("CANVUS_BASE_URL")
	if err != nil {
		return err
	}
	apiKey, err := mustEnv("CANVUS_API_KEY")
	if err != nil {
		return err
	}
	canvasID, err := mustEnv("CANVUS_CANVAS_ID")
	if err != nil {
		return err
	}

	cfg := &canvus.SessionConfig{
		BaseURL: baseURL,
		// Long timeout — the stream itself is long-lived. The SDK's default
		// 30s timeout would kill the request after 30s; we want minutes.
		RequestTimeout: 10 * time.Minute,
	}
	s := canvus.NewSession(cfg, canvus.WithAPIKey(apiKey))

	duration := streamDuration()
	slog.Info("starting subscribe",
		"canvas_id", canvasID,
		"duration_seconds", int(duration.Seconds()))

	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer cancel()

	deadline, cancelDeadline := context.WithTimeout(ctx, duration)
	defer cancelDeadline()

	stats, err := stream(deadline, s, canvasID)
	if err != nil {
		// Cancellation is not a failure — it's our exit condition.
		if ctxErr := deadline.Err(); ctxErr != nil {
			slog.Info("stream ended on context",
				"reason", ctxErr,
				"frames", stats.frames,
				"non_empty_frames", stats.nonEmpty)
			return nil
		}
		return fmt.Errorf("stream: %w", err)
	}
	slog.Info("stream completed", "frames", stats.frames, "non_empty_frames", stats.nonEmpty)
	return nil
}

type streamStats struct {
	frames   int64
	nonEmpty int64
}

// stream issues the subscribe request and prints one summary line per
// non-empty frame.
func stream(ctx context.Context, s *canvus.Session, canvasID string) (streamStats, error) {
	var stats streamStats

	u, err := url.Parse(s.BaseURL)
	if err != nil {
		return stats, fmt.Errorf("parse base URL: %w", err)
	}
	u.Path = path.Join(u.Path, fmt.Sprintf("canvases/%s/notes", canvasID))
	q := u.Query()
	q.Set("subscribe", "true")
	u.RawQuery = q.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return stats, fmt.Errorf("new request: %w", err)
	}
	req.Header.Set("Accept", "application/x-ndjson, application/json")
	// Authentication is added by the round-tripper installed by WithAPIKey.

	resp, err := s.HTTPClient.Do(req)
	if err != nil {
		return stats, fmt.Errorf("do: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
		return stats, fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(body))
	}

	scanner := bufio.NewScanner(resp.Body)
	// Notes responses can be many KB for busy canvases.
	scanner.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)

	for scanner.Scan() {
		line := scanner.Bytes()
		atomic.AddInt64(&stats.frames, 1)
		if len(line) == 0 {
			// Keep-alive frame per the SDK observation; skip silently.
			continue
		}
		atomic.AddInt64(&stats.nonEmpty, 1)
		printFrame(line)
	}
	if err := scanner.Err(); err != nil {
		return stats, fmt.Errorf("scan: %w", err)
	}
	return stats, nil
}

// printFrame decodes a single NDJSON frame and prints a one-line summary.
// The server emits arrays of notes; if the shape ever diverges we still
// print the raw byte count so the caller sees activity.
func printFrame(line []byte) {
	var notes []canvus.Note
	if err := json.Unmarshal(line, &notes); err != nil {
		slog.Warn("frame did not parse as []Note", "err", err, "bytes", len(line))
		return
	}
	var firstID, firstText string
	if len(notes) > 0 {
		firstID = notes[0].ID
		firstText = truncate(notes[0].Text, 60)
	}
	slog.Info("frame",
		"ts", time.Now().UTC().Format(time.RFC3339Nano),
		"count", len(notes),
		"first_id", firstID,
		"first_text", firstText)
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

func streamDuration() time.Duration {
	if v := os.Getenv("STREAM_DURATION_SECONDS"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			return time.Duration(n) * time.Second
		}
		slog.Warn("STREAM_DURATION_SECONDS could not be parsed; using default",
			"value", v, "default_seconds", defaultDurationSeconds)
	}
	return defaultDurationSeconds * time.Second
}

// mustEnv returns the value of name or an error if it is empty/unset.
func mustEnv(name string) (string, error) {
	v := os.Getenv(name)
	if v == "" {
		return "", fmt.Errorf("missing required env var %s", name)
	}
	return v, nil
}

// setupLogging configures slog.Default.
func setupLogging() {
	var h slog.Handler
	switch os.Getenv("LOG_FORMAT") {
	case "json":
		h = slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	default:
		h = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	}
	slog.SetDefault(slog.New(h))
}
